readthedocs/core/views/__init__.py

```diff
@@ -124,7 +129,36 @@ def server_error_404(request, exception=None, template_name='404.html'):  # pyli
             )
         else:
             return response
-    r = render(request, template_name)
+
+    project, path = project_and_path_from_request(request, request.get_full_path())
```

Comment:
This function comes from `readthedocs.redirects.utils` but we are not using it here for any redirect. This may worth a refactor since this function is called here and also inside of `get_redirect_response` some lines above.